endpoint/platform/simulator/simulator.c

```diff
@@ -0,0 +1,149 @@
+  if (err != SC_OK) LE_ERROR("register simulator device error:%d", err);
+}
+
+static inline void unregister_simulator(void) {
+  status_t ret = unregister_device(&simulator_device_type);
+  LE_INFO("unregister simulator return: %d", ret);
+}
+
+static status_t simulator_init(void) {
+  register_simulator();
+  return SC_OK;
+}
+
+static void simulator_release(void) { unregister_simulator(); }
+
+static status_t simulator_get_key(uint8_t *key) {
+  memcpy(key, private_key, 16);
+  LE_INFO("Get device key success");
+  return SC_OK;
+}
+
+static status_t simulator_get_device_id(char *id) {
+  memcpy(id, device_id, 16);
+  LE_INFO("Get device id success");
+  return SC_OK;
+}
+
+static status_t uart_init(const char *device) {
+  LE_INFO("UART init device %s success", device);
+  return SC_OK;
+}
+
+static void uart_write(const int fd, const char *cmd) {
+  /* simple output */
+  size_t cmd_len = strlen(cmd);
+  if (cmd_len > UART_BUFFER_SIZE) {
+    LE_ERROR("command too long");
+    return;
+  }
+  snprintf(UART_BUFFER, cmd_len, "%s", cmd);
+  LE_INFO("UART write success");
+}
+
+static char *uart_read(const int fd) {
+  char *response = strdup("This is a test");
+  LE_INFO("UART read success");
+  return response;
+}
+
+static void uart_clean(const int fd) { LE_INFO("UART clean success"); }
+
+static status_t sec_init(void) {
+  LE_INFO("Initialize secure storage");
+  simulator_sec_table =
+      le_hashmap_Create("Simulator secure storage", 32, le_hashmap_HashString, le_hashmap_EqualsString);
+  return SC_OK;
+}
+
+static status_t sec_write(const char *name, const uint8_t *buf, size_t buf_size) {
+  uint8_t *data = malloc(buf_size);
+  uint8_t *ptr;
+  memcpy(data, buf, buf_size);
+
+  LE_INFO("Write %s into secure storage", name);
+
+  // the hashmap will return old value if it is replaced
+  ptr = le_hashmap_Put(simulator_sec_table, name, data);
+  if (ptr != NULL) {
+    free(ptr);
+  }
+  return SC_OK;
+}
+
+static status_t sec_read(const char *name, uint8_t *buf, size_t *buf_size) {
+  LE_INFO("Read %s from secure storage", name);
+  uint8_t *data = le_hashmap_Get(simulator_sec_table, name);
+
+  if (data == NULL) {
+    *buf_size = 0;
+    return SC_ENDPOINT_SEC_ITEM_NOT_FOUND;
+  }
+
+  memcpy(buf, data, *buf_size);
+  return SC_OK;
+}
+
+static status_t sec_delete(const char *name) {
+  LE_INFO("Delete %s in secure storage", name);
+  uint8_t *data = le_hashmap_Remove(simulator_sec_table, name);
+
+  if (data == NULL) {
+    return SC_ENDPOINT_SEC_ITEM_NOT_FOUND;
+  }
+
+  free(data);
+  return SC_OK;
+}
+
+static const struct device_operations simulator_ops = {.init = simulator_init,
+                                                       .fini = simulator_release,
+                                                       .get_key = simulator_get_key,
+                                                       .get_device_id = simulator_get_device_id};
+
+static const struct uart_operations simulator_uart = {
```

Comment:
Style tweaks:
* Split into several new lines for each item.
* Append `,` (comma) right after the last item to satisfy clang-format.